crates/arroyo-operator/src/operator.rs

```diff
@@ -232,20 +301,38 @@ async fn operator_run_behavior(
                                 ).await;
                             }
                             ArrowMessage::Signal(signal) => {
-                                match this.handle_control_message(idx, &signal, &mut counter, &mut closed, in_partitions, ctx).await {
+                                match this.handle_control_message(idx, &signal, &mut counter, &mut closed,&mut finished, in_partitions, ctx).await {
                                     ControlOutcome::Continue => {}
-                                    ControlOutcome::Stop => {
-                                        // just stop; the stop will have already been broadcast for example by
-                                        // a final checkpoint
-                                        break;
+                                    ControlOutcome::FinishedStoppingCheckpoint => {
+                                        info!("finished stopping checkpoint in operator {}-{}, waiting for terminate",
+                                            ctx.task_info.operator_name, ctx.task_info.task_index);
+                                        operator_state = OperatorState::WaitingForTerminate;
                                     }
-                                    ControlOutcome::Finish => {
-                                        final_message = Some(SignalMessage::EndOfData);
-                                        break;
+                                    ControlOutcome::NoMoreData { end_of_data } => {
+                                        if operator_state != OperatorState::Running {
+                                            warn!("received no more data update in operator {}-{} while in state {:?}",
```

Comment:
I consistently see this warning when stopping pipelines — is it an actual issue, or should we make it a debug?